src/main/java/com/google/devtools/build/lib/rules/cpp/LibrariesToLinkCollector.java

```diff
@@ -236,20 +240,26 @@ public CollectedLibrariesToLink collectLibrariesToLink() {
       NestedSetBuilder<LinkerInput> expandedLinkerInputsBuilder) {
     boolean includeSolibDir = false;
     boolean includeToolchainLibrariesSolibDir = false;
+    HashMap<String, PathFragment> linkedLibrariesPaths = new HashMap<>();
+
     for (LinkerInput input : linkerInputs) {
       if (input.getArtifactCategory() == ArtifactCategory.DYNAMIC_LIBRARY
           || input.getArtifactCategory() == ArtifactCategory.INTERFACE_LIBRARY) {
         PathFragment libDir = input.getArtifact().getExecPath().getParentDirectory();
+        String libraryIdentifier = input.getLibraryIdentifier();
+        PathFragment previousLibDir = linkedLibrariesPaths.get(libraryIdentifier);
+        if (previousLibDir == null) {
+          linkedLibrariesPaths.put(libraryIdentifier, libDir);
+        } else if (!previousLibDir.equals(libDir)) {
+          ruleErrorConsumer.ruleError(
+              String.format(
+                  "You are trying to link the same dynamic library %s built in a different configuration",
```

Comment:
I think you are can give more information in the error by giving the artifact path of the two libraries that you want to link. The path should have a configuration segment which people can use to debug.